imgix/react-imgix
# feat: implement responsiveness with srcSet and sizes

BREAKING CHANGE: 
- srcSet behaviour has changed to use sizes + srcSets
- `type=bg` has been removed
- the following props have been removed: aggressiveLoad, component, fluid, precision, defaultHeight, defaultWidth
- generateSrcSet has been changed to disableSrcSet


## Description

This PR fundamentally changes the approach taken to responsiveness. Previously, the image (or component) was rendered with an empty image, and the size of the resulting DOM element was measured. An image was loaded with these dimensions, ensuring that the image loaded was only as large as needed. Now, this component achieves the same behaviour by leveraging new native browser features, `srcSet` and `sizes`. This allows the browser to load an appropriate size based on the expected image size, which is calculated based on `sizes`. This has an added benefit that images can load much quicker as the browser can load them as soon as page layout has completed, or sooner (e.g. if sizes is `100vw`). Overall this change has made both the usage of the library and the internals simpler and easier to understand.

The significant amount of breaking changes makes sense due to a lot of the props no longer making sense. Background mode has also been removed as it was incompatible with the new method of responsiveness, and after discussion with @jayeb it seemed that it wouldn't be missed (based on a similar change that imgix.js made).

Picture mode still doesn't work amazingly, as there is a mis-match between what a developer would expect to set as props and what they actually have to set to make picture work as they want. This is not a problem with our library per se, but with `picture` pecularities. For example, if the fallback `img` in a `<picture>` has a `width` attribute set, *all* the other sources will also be locked to that size, even if they specify their own widths. Maybe this is common knowledge but it seems peculiar to me. I expect to address this and make it clearer for developers in a future PR.

This PR closes #153. This PR closes #115. This PR closes #71

## Checklist

Please use the checklist that is most closely related to your PR _(you only need to use one checklist, and you can skip items that aren't applicable or don't make sense)_:

- [fixing typos]()
- [documentation]()
- [bug fix]()
- [new feature]()

### New Feature

- [x] If this is a big feature with breaking changes, consider [opening an issue][issues] to discuss first. This is completely up to you, but please keep in mind that your PR might not be accepted.

## Steps to Test

Observe changes to unit tests. 

It may also be worth to set up a test project and use `npm link` to check default image rendering and picture rendering.
